### DISCUSSION TITLE
Displaying the national font

### DISCUSSION THREAD

**Discussion by Keyn1** [raysan5/raylib#2499] [Q&A] [answered]
Tell me, I use the golang+raylab binding. The national font is not displayed. Prints like this: "____ ________" What needs to be done to display the national font?

I took the program from the example
	// TTF Font loading with custom generation parameters
	font := rl.LoadFontEx("KAISG.ttf", 250, nil, 0) // 96

	// Generate mipmap levels to use trilinear filtering
	// NOTE: On 2D drawing it won't be noticeable, it looks like FILTER_BILINEAR
	rl.GenTextureMipmaps(&font.Texture)
...
rl.DrawText("Use mouse wheel to change font size По русски пишешь?", 20, 20, 10, rl.Gray)

**Comment by JupiterRider** [comment]
Hey Keyn1,

seems to be an issue with raylib and ttf fonts. Can you try using a BMFont?
You can use tools to generate one:
[https://libgdx.com/wiki/tools/hiero](https://libgdx.com/wiki/tools/hiero)
[https://www.angelcode.com/products/bmfont/](https://www.angelcode.com/products/bmfont/)

Cheers

**Comment by JupiterRider** [comment, marked as answer]
@raysan5 
Do you know something about raylib's ttf support for Cyrillic characters?
Doesn't seem to work with the default font as well.

**Reply by raysan5** [reply]
The generated font atlas should include the required glyphs. 

The list with required characters should be provided to the function `LoadFontEx()`:
```c
Font LoadFontEx(const char *fileName, int fontSize, int *fontChars, int glyphCount);
``` 
`fontChars` is a list of integer codepoints, every codepoint represents one glyph, Cyrillic glyphs are represented by codepoints 0x400 to 0x4ff values. Here some sample code:
```c
int codepoints[512] = { 0 };
for (int i = 0; i < 95; i++) codepoints[i] = 32 + i;   // Basic ASCII characters
for (int i = 0; i < 255; i++) codepoints[96 + i] = 0x400 + i;   // Cyrillic characters
Font font = LoadFontEx("font_containing_cyrillic_glyphs.ttf", 32, codepoints, 512);
```
I think it should work.

The default font does not include Cyrillic glyphs.

**Reply by JupiterRider** [reply]
It does work!

**Reply by DesLandysh** [reply]
Same for **python** users:

```python
# define list of integers that represents the glyphs: ASCII and Cyrillic
codepoints = [0x380 + i if i > 127 else i for i in range(32, 512)]

# to get some special symbols and full Greek alphabet, list size will be exact 512
# codepoints = [0x240 + i if i > 127 else i for i in range(32, 544)]
```

and use this list in font loading extension function:
```python
my_font = load_font_ex("PATH_TO_FONT.ttf", 32, codepoints, 512)
draw_text_ex(my_font, "Съешь еще этих мягких булок и выпей чаю!:;?.,%&@'\"#№*/-+()", (x_pos, y_pos), font_size, spacing, COLOR)
draw_text_ex(my_font, "Brown fox jumps over the lazy dog!:;?.,%&@'\"#№*/-+()", (x_pos, y_pos), font_size, spacing, COLOR)
```

**Reply by Hammer2900** [reply]
char_codes = [0x0380 + i if i > 127 else i for i in range(32, 512)]
font_chars = pyray.ffi.new('int[]', char_codes)
font_chars_ptr = pyray.ffi.cast('int *', font_chars)
glyph_count = len(char_codes)
font = pyray.load_font_ex(font_path.encode('utf-8'), font_size, font_chars_ptr, glyph_count)
